dockstore-webservice/src/main/java/io/dockstore/webservice/core/Entry.java

```diff
@@ -249,6 +252,16 @@
     @ApiModelProperty(value = "The presence of the put code indicates the entry was exported to ORCID.")
     private String orcidPutCode;
 
+    /**
+     * A set of this Entry's Collections from which we can filter Categories.
+     * Don't try to directly manipulate this field, bad things will happen.
```

Comment:
Could you elaborate on this (bad things)?
In context, since this doesn't actually add any tables to the DB, it seems what this accomplishes could be done with named or native queries which might be better than leaving behind a field that could cause bad things if used?

Reply:
If you call add() or remove() on this field, the system will try to update the db, which is not desirable.
The rationale for the above field is to avoid having to run a Category query deep in the Elasticsearch indexing code, because getting the necessary data structures to the point of invocation was difficult and very unclean.  If we could just call a method in Entry which would run such a query, that would work, but how?  There's no session object in Entry.

Reply:
Oh, in the past if we didn't want a property to be able to make permanent changes. We would make it `@Transient` and conditionally populated from a resource method. Examples are `Tool.setTags` or `Collection.setCollectionEntries`

I think would help with the funny business with `import.sql` as well.

Reply:
Yeah, I understand the "stuff-it-into-a-transient" approach.  Generally, the current code uses this technique on-demand, to populate fields that are going to be used immediately afterwards.  In this case, the categories of an Entry need to be populated before it hits the ES update code, to which it is passed via handleIndexUpdate().  There's 20+ invocations of handleIndexUpdate() scattered about, so "stuffing-it-into-a-transient" prior to that call is probably not recommended.  If we "stuff" after that point, we need to propagate the necessary state into PublicStateManager/etc to do the categories query, and the changes required were ugly (I tried it).

Reply:
Certainly, we could create another function populateAndIndex(Entry) which "stuffs-the-transient" then invokes handleIndexUpdate(), and change all invocations of handleIndexUpdate() to populateAndIndex().  It's a bigger change, and it's easy to call mistakenly call handleIndexUpdate() directly, but might be cleaner, depending on who you talk to.

Reply:
Just took another look at the code, and I think there's a way to "stuff-the-transient" in each to-be-indexed Entry using a StateListener that is called before the ElasticListener.  I may need to open up the PublicStateManager initialization process to make it work.  Will consult with Denis on Monday.